File: core/src/si/feri/slidegame/screen/map/Markers.java
package si.feri.slidegame.screen.map;


import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.scenes.scene2d.Group;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import si.feri.slidegame.config.GameConfig;
import si.feri.slidegame.database.Database;
import si.feri.slidegame.database.Event;
import si.feri.slidegame.database.EventFirebase;
import si.feri.slidegame.utils.Geolocation;
import si.feri.slidegame.utils.Map;
import si.feri.slidegame.utils.PixelPosition;

import java.io.IOException;
import java.util.HashMap;
import java.util.UUID;

public class Markers extends Group {
    public final Map map;
    public final SideMenue sideMenue;
    public final TextureRegion markerRegion;


    public Markers(MapScreen mapScreen, TextureRegion markerRegion) {
        this.map = mapScreen.map;
        this.markerRegion = markerRegion;
        this.sideMenue = mapScreen.sideMenue;
    }

    //
    // Refresh
    //

    public static float REFRESH_INTERVAL = 10f;
    private float refreshTime = 0;

    public final HashMap<UUID, Marker> markers = new HashMap<>();

    @Override
    public void act(float delta) {
        super.act(delta);

        //
        //
        //

        refreshTime -= delta;
        if (refreshTime <= 0) {
            refreshTime = REFRESH_INTERVAL;
            try {
                for (final EventFirebase firEvent : Database.fetchEvents().values()) {
                    Event event;
                    try {
                        event = firEvent.parse();
                    } catch (Exception e) {
                        System.out.println(e);
                        System.out.println(firEvent);
                        event = new Event();
                    }

                    Marker marker = markers.get(event.id);
                    if (marker != null) {
                        if(marker.updated || marker == sideMenue.marker) continue;
                        marker.update(event);
                    } else {
                        Marker newMarker = new Marker(event);
                        markers.put(event.id, newMarker);
                        addActor(newMarker);
                    }
                }
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

    }

    //
    //
    //

    public class Marker extends Image {
        public boolean updated = false;
        public Event event;
        public PixelPosition pixelPosition;

        public Marker(Event event) {
            super(markerRegion);
            setWidth(GameConfig.MARKER_WIDTH);
            setHeight(GameConfig.MARKER_HEIGHT);
            update(event);

            //
            addListener(new ClickListener() {
                @Override
                public void clicked(InputEvent event, float x, float y) {
                    sideMenue.setMarker(Marker.this);
                }
            });
        }

        public void update(Event event) {
            this.event = event;
            recomputePosition();
        }

        public void recomputePosition() {
            pixelPosition = map.getPixelPosition(event.location);
            setPosition(pixelPosition.x, pixelPosition.y);
        }


        public void setLocation(Geolocation geo) {
            event.location = geo;
            pixelPosition = map.getPixelPosition(event.location);
            setPosition(pixelPosition.x - getWidth()/2, pixelPosition.y);
        }
    }
}

File: core/src/si/feri/rrizemljevid/screen/map/Markers.kt
package si.feri.rrizemljevid.screen.map

import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.scenes.scene2d.Group
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import si.feri.rrizemljevid.config.GameConfig
import si.feri.rrizemljevid.database.Database
import si.feri.rrizemljevid.database.Event
import si.feri.rrizemljevid.utils.Geolocation
import si.feri.rrizemljevid.utils.Map
import si.feri.rrizemljevid.utils.PixelPosition
import java.io.IOException
import java.util.*

class Markers(mapScreen: MapScreen, val markerRegion: TextureRegion) : Group() {
    val map: Map?
    val sideMenue: SideMenue?
    private var refreshTime = 0f
    val markers = HashMap<UUID?, Marker>()

    init {
        map = mapScreen.map
        sideMenue = mapScreen.sideMenue
    }

    override fun act(delta: Float) {
        super.act(delta)

        //
        //
        //
        refreshTime -= delta
        if (refreshTime <= 0) {
            refreshTime = REFRESH_INTERVAL
            try {
                for (firEvent in Database.fetchEvents().values) {
                    var event: Event?
                    event = try {
                        firEvent.parse()
                    } catch (e: Exception) {
                        println(e)
                        println(firEvent)
                        Event()
                    }
                    val marker = markers[event!!.id]
                    if (marker != null) {
                        if (marker.updated || marker === sideMenue!!.marker) continue
                        marker.update(event)
                    } else {
                        val newMarker: Marker = Marker(event)
                        markers[event.id] = newMarker
                        addActor(newMarker)
                    }
                }
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
    }

    //
    //
    //
    inner class Marker(event: Event?) : Image(
        markerRegion
    ) {
        var updated = false
        var event: Event? = null
        var pixelPosition: PixelPosition? = null

        init {
            width = GameConfig.MARKER_WIDTH
            height = GameConfig.MARKER_HEIGHT
            update(event)

            //
            addListener(object : ClickListener() {
                override fun clicked(event: InputEvent, x: Float, y: Float) {
                    sideMenue!!.marker = this@Marker
                }
            })
        }

        fun update(event: Event?) {
            this.event = event
            recomputePosition()
        }

        fun recomputePosition() {
            pixelPosition = map!!.getPixelPosition(event!!.location)
            setPosition(pixelPosition!!.x.toFloat(), pixelPosition!!.y.toFloat())
        }

        fun setLocation(geo: Geolocation?) {
            event!!.location = geo!!
            pixelPosition = map!!.getPixelPosition(event!!.location)
            setPosition(pixelPosition!!.x - width / 2, pixelPosition!!.y.toFloat())
        }
    }

    companion object {
        //
        // Refresh
        //
        var REFRESH_INTERVAL = 10f
    }
}